apollo-test/src/main/java/com/spotify/apollo/test/StubClient.java

```diff
@@ -275,13 +319,48 @@ private ResponseSource responseSource() {
 
   }
 
+
+  public class WhenResponseBuilder {
+
+    private final Matcher<Request> requestMatcher;
+
+    public WhenResponseBuilder(Matcher<Request> requestMatcher) {
+      this.requestMatcher = Objects.requireNonNull(requestMatcher);
+    }
+
+    /**
+     * Immediately respond with the supplied response to any matching request.
+     */
+    public void respond(Response<ByteString> response) {
+      addMatcherFirst(requestMatcher, Responses.constant(ResponseWithDelay.forResponse(response)));
+    }
+
+    /**
+     * Respond with responses and delays supplied by the provided ResponseSource to any matching
+     * request.
+     */
+    public void respond(ResponseSource responseSource) {
+      addMatcherFirst(requestMatcher, responseSource);
+    }
+  }
+
   @AutoValue
-  public static abstract class RequestResponsePair {
+  static abstract class RequestResponsePair {
     public abstract Request request();
     public abstract Response<ByteString> response();
 
     public static RequestResponsePair create(Request request, Response<ByteString> response) {
       return new AutoValue_StubClient_RequestResponsePair(request, response);
     }
   }
+
+  @AutoValue
+  static abstract class MatcherResponseSourcePair {
```

Comment:
I think a private static final class with final members would cause the least cognitive overhead in this case.